File: 282-expression-add-operators/expression-add-operators.py
class Solution:
    def addOperators(self, num: str, target: int) -> List[str]:
        result = []
        n = len(num)
        
        def backtrack(index: int, prev_op: int, curr_val: int, path: List[str]):
            if index == n:
                if curr_val == target:
                    result.append("".join(path))
                return
                
            for j in range(index, n):
                if j > index and num[index] == '0':
                    break
                    
                sub_str = num[index:j+1]
                val = int(sub_str)
                
                if index == 0:
                    path.append(sub_str)
                    backtrack(j + 1, val, val, path)
                    path.pop()
                else:
                    path.append('+')
                    path.append(sub_str)
                    backtrack(j + 1, val, curr_val + val, path)
                    path.pop()
                    path.pop()
                    
                    path.append('-')
                    path.append(sub_str)
                    backtrack(j + 1, -val, curr_val - val, path)
                    path.pop()
                    path.pop()
                    
                    path.append('*')
                    path.append(sub_str)
                    backtrack(j + 1, prev_op * val, (curr_val - prev_op) + (prev_op * val), path)
                    path.pop()
                    path.pop()
                    
        backtrack(0, 0, 0, [])
        return result

File: 282-expression-add-operators/expression-add-operators.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<String> addOperators(String num, int target) {
        List<String> result = new ArrayList<>();
        if (num == null || num.length() == 0) {
            return result;
        }
        backtrack(result, new StringBuilder(), num, target, 0, 0, 0);
        return result;
    }

    private void backtrack(List<String> result, StringBuilder sb, String num, int target, int index, long eval, long prev) {
        if (index == num.length()) {
            if (eval == target) {
                result.add(sb.toString());
            }
            return;
        }

        for (int i = index; i < num.length(); i++) {
           
            if (i != index && num.charAt(index) == '0') {
                break;
            }

            long curr = Long.parseLong(num.substring(index, i + 1));
            int len = sb.length();

            if (index == 0) {
                sb.append(curr);
                backtrack(result, sb, num, target, i + 1, curr, curr);
                sb.setLength(len);
            } else {
               
                sb.append("+").append(curr);
                backtrack(result, sb, num, target, i + 1, eval + curr, curr);
                sb.setLength(len);

                
                sb.append("-").append(curr);
                backtrack(result, sb, num, target, i + 1, eval - curr, -curr);
                sb.setLength(len);

               
                sb.append("*").append(curr);
                backtrack(result, sb, num, target, i + 1, eval - prev + (prev * curr), prev * curr);
                sb.setLength(len);
            }
        }
    }
}
